### ISSUE TITLE
weird output since 7094a82f2ea7dbc2f2a794c846bb540526c7ee00

### ISSUE THREAD

**Issue by koenokatachi** [biometrics/openbr#426]
7094a82f2ea7dbc2f2a794c846bb540526c7ee00 removes ```biometrics/openbr-models``` from submodules and puts this to cmake file instead

```
ExternalProject_Add(models
                    URL http://github.com/biometrics/openbr/releases/download/v1.0.0/models.tar.gz
...
```

but the tar doesnt include transforms found in ```biometrics/openbr-models``` (FaceRecognitionQuantization, FaceRecognitionExtraction, FaceRecognitionEmbedding).

pre 7094a82f2ea7dbc2f2a794c846bb540526c7ee00:
br -algorithm FaceRecognition -compare 1.png 1.png
7.14759
br -algorithm FaceRecognition -compare 1.png 1_half.png
4.95547
br -algorithm FaceRecognition -compare 1.png 2.png
0.00889185

post 7094a82f2ea7dbc2f2a794c846bb540526c7ee00 (built with http://openbiometrics.org/docs/install/index.html#linux):
br -algorithm FaceRecognition -compare 1.png 1.png
0
br -algorithm FaceRecognition -compare 1.png 1_half.png
-0.517358
br -algorithm FaceRecognition -compare 1.png 2.png
-0.885855

samples:
![1](https://cloud.githubusercontent.com/assets/11669814/9903645/b7f5116e-5c80-11e5-8681-bbe0cf7d7d4c.png)

![1_half](https://cloud.githubusercontent.com/assets/11669814/9903650/bd2477ec-5c80-11e5-8cb0-d64ed22b9112.png)

![2](https://cloud.githubusercontent.com/assets/11669814/9903652/c03a6234-5c80-11e5-81a7-dbf3aaffe65c.png)

**Comment by jklontz** [comment]
@bklare Can you add to your todo list retraining with score normalization? @koenokatachi For the time being with the new algorithm values closer to zero indicate greater similarity.

**Comment by koenokatachi** [comment]
here are some pictures

org.png
![org](https://cloud.githubusercontent.com/assets/11669814/9936812/7d2fdf28-5d66-11e5-8a3f-2618f52e28e8.png)

crop.png (same org.png but cropped. the face is intact)
![crop](https://cloud.githubusercontent.com/assets/11669814/9936835/91b24634-5d66-11e5-8771-0a3d5b754402.png)

chop-1px-left.png (org.png with 1px chopped from left)
![chop-1px-left](https://cloud.githubusercontent.com/assets/11669814/9936854/a8d3d576-5d66-11e5-8da9-2cef48be0559.png)

chop-1px-right.png (org.png with 1px chopped from right)
![chop-1px-right](https://cloud.githubusercontent.com/assets/11669814/9936891/cff5a1fc-5d66-11e5-9d42-1947a427e51e.png)

all pictures feature the same face.

and now have a look at the numbers:
```
br -algorithm FaceRecognition -compare org.png crop.png 
Set algorithm to FaceRecognition
Loading /usr/local/share/openbr/models/algorithms/FaceRecognition
Comparing org.png and crop.png
Enrolling org.png to orgRxJIl2.mem
100.00%  ELAPSED=00:00:00  REMAINING=00:00:00  COUNT=1
100.00%  ELAPSED=00:00:00  REMAINING=00:00:00  COUNT=1
-0.52921

br -algorithm FaceRecognition -compare org.png chop-1px-left.png 
Set algorithm to FaceRecognition
Loading /usr/local/share/openbr/models/algorithms/FaceRecognition
Comparing org.png and chop-1px-left.png
Enrolling org.png to orgRxJIl2.mem
100.00%  ELAPSED=00:00:00  REMAINING=00:00:00  COUNT=1
100.00%  ELAPSED=00:00:00  REMAINING=00:00:00  COUNT=1
-0.437604

br -algorithm FaceRecognition -compare org.png chop-1px-right.png 
Set algorithm to FaceRecognition
Loading /usr/local/share/openbr/models/algorithms/FaceRecognition
Comparing org.png and chop-1px-right.png
Enrolling org.png to orgRxJIl2.mem
100.00%  ELAPSED=00:00:00  REMAINING=00:00:00  COUNT=1
100.00%  ELAPSED=00:00:00  REMAINING=00:00:00  COUNT=1
0

br -algorithm FaceRecognition -compare chop-1px-*
Set algorithm to FaceRecognition
Loading /usr/local/share/openbr/models/algorithms/FaceRecognition
Comparing chop-1px-left.png and chop-1px-right.png
Enrolling chop-1px-left.png to chop-1px-leftvKaoYj.mem
100.00%  ELAPSED=00:00:00  REMAINING=00:00:00  COUNT=1
100.00%  ELAPSED=00:00:00  REMAINING=00:00:00  COUNT=1
-0.437604
```

and for completely different faces it gives -0.885855 (Obama vs Nina in my first comment)

how am i supposed to recognize faces with such output? it gives ~-0.5 for the same faces and ~-0.9 for completely different ones. how do i interpret that?

with the pre-commit build i get such numbers:
5.23928
5.0709
7.14759
5.0709

and 0.00889185 for Obama vs Nina.

it is pretty consistent and solid compare to what you called the new algo values

**Comment by koenokatachi** [comment]
@jklontz  Btw could you please approve the problem isnt by my side? Can you reproduce the scores with the master build?

I've tried it in 3 instances: my local machine with ubuntu 12.04 and 14.04, virtual remote server with 14.04. All the same results.

I've even trained the algorithm on the LWF set (it took 2h 20m). And I've got almost the same scores when doing -compare with the newly trained model.

And also what is the point of using negLogPlusOne=true by default (pretrained FaceRecognition uses Dist(L2) which turns into Dist(L2,negLogPlusOne=true) because of defaults)? It is dissapointing getting negative measure value.

**Comment by jklontz** [comment]
Strictly speaking this isn't a bug, it's just that we forgot to ship the algorithm with score normalization that would linearly map the raw negative scores you are seeing into the positive range akin to the old algorithm. I'll bug @bklare to fix this when he gets back into town. Until then you'll just have to use a threshold of around `-0.8` (plus or minus a bit depending on your tolerance for false positives vs. false negatives).

**Comment by koenokatachi** [comment]
Looking at the algo's definition. It ends with ```...+LDA(768)+Normalize(L2):Dist(L2)```

So when comparing images ```Dist(L2)``` is done over two unit L2 length vectors so the output (when it works as supposed) WITHOUT applying ```-ln(1+x)``` is gonna be in ```[0,2]```?

Btw check out this line https://github.com/biometrics/openbr/blob/master/openbr/plugins/classification/lda.cpp#L427 mby it is the source of the problem. If the mean value of ```results``` is not 0 then ```stdDev``` is not the actual std deviation.

**Comment by marcbackes** [comment]
I saw you updated this today (https://github.com/biometrics/openbr/commit/9633db66d3958cbc5607a59904c49f81676619a1).

But it still isn't working. I still get -0.88... and similar values. For re-installing, do I have to do anything in particular? I just re-cloned the repository and installed the software as usual.

Best regards,
Marc

**Comment by jklontz** [comment]
We haven't added the new models yet, will close this issue when it's resolved :)

**Comment by marcbackes** [comment]
Thanks for the quick response. Is there any workaround I can do as a quick fix? Maybe copy the files in the source and rebuild it? The thing is, I use openbiometrics in a project and on Monday I have to present it ;)

Also, getting version 0.5 to work could be an option.

Best regards.

**Comment by bklare** [comment]
There is a tagged release for version 1.0. I would not be working off of
master unless you need some specific update.

**Comment by koenokatachi** [comment]
@bklare tagged 1.0 has the same problem

**Comment by jklontz** [comment]
Fixed in 16a3b3a7dc4dc12bb9e359d8ba9b2edf9adc471d
